ci_cd/utils.py

```diff
@@ -65,14 +68,45 @@ class SemanticVersion:
 
     """
 
+    _REGEX = (
+        r"^(?P<major>0|[1-9]\d*)(?:\.(?P<minor>0|[1-9]\d*))?(?:\.(?P<patch>0|[1-9]\d*))?"
+        r"(?:-(?P<pre_release>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)"
+        r"(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
+        r"(?:\+(?P<build>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"
+    )
+
     def __init__(
         self,
-        major: "Union[str, int]",
+        version: "Optional[str]" = None,
+        major: "Optional[Union[str, int]]" = None,
         minor: "Optional[Union[str, int]]" = None,
         patch: "Optional[Union[str, int]]" = None,
         pre_release: "Optional[str]" = None,
         build: "Optional[str]" = None,
     ) -> None:
+        if version:
+            if major or minor or patch or pre_release or build:
+                raise ValueError(
+                    "version cannot be specified along with other parameters"
+                )
+
+            if not isinstance(version, str):
```

Comment:
Mnjeh. Yes and no.
So pydantic indeed does type validation for any pydantic models.
`SemanticVersion` is _not_ a pydantic model, but rather a "standard" Python class. Actually, it's a subclass of the built-in `str` type.

Furthermore, it's always good practice to verify/validate types of variables in some function or class (if this is not automagically done for you as is done by pydantic models), _even if_ you know that you're only passing in `str` or similar types to some function or class.
That's a bit the point of modularization, to have modules interact with each other "independently" of each other, so that if one gets the interface (input/outputs) right for other functions or classes, once can simply drop-in and drop-out different modules and parts that make up a whole.
Hence, each individual module (in this case some function of class) should self-validate that the variables it receives are what it expects and not trust that it's just correct.
Again - if the class here was a pydantic model, the model does the validation for me, so I wouldn't have to, but as this is _not_ a pydantic model, I have to do a modicum of validation myself.

Reply:
Very good. Thank you!